back/config/initializers/rack_attack.rb

```diff
@@ -31,13 +31,25 @@ class Rack::Attack
     end
   end
 
+  throttle('logins/ip/day', limit: 4000, period: 24.hours) do |req|
+    if req.path == '/web_api/v1/user_token' && req.post?
+      req.remote_ip
+    end
+  end
+
   # Signing in by email account.
   throttle('logins/email', limit: 10, period: 20.seconds) do |req|
     if req.path == '/web_api/v1/user_token' && req.post?
       JSON.parse(req.body.string).dig('auth', 'email')&.to_s&.downcase&.gsub(/\s+/, "")&.presence
     end
   end
 
+  throttle('logins/email/day', limit: 100, period: 24.hours) do |req|
+    if req.path == '/web_api/v1/user_token' && req.post?
+      JSON.parse(req.body.string).dig('auth', 'email')&.to_s&.downcase&.gsub(/\s+/, "")&.presence
```

Comment:
This will throw an error if the request body does not contain a valid JSON document. What do you think about making the parsing more robust and prevent a 500 from happening?

Reply:
Sounds like a good idea, but I'm not sure how to do this. I have tried...
```
begin
  JSON.parse(req.body.string).dig('auth', 'email')&.to_s&.downcase&.gsub(/\s+/, "")&.presence
rescue JSON::ParserError
  puts "JSON::ParserError"
end
```
But I still get error: `JSON::ParserError: 859: unexpected token at 'INVALID JSON'`

Reply:
Mh Simon, this works for me on my machine, some hints:

- Did you add the rescue clause to both `throttle` blocks that check the `'/web_api/v1/user_token'` path?
- Did you restart the web container? I'm not sure if Rails automatically detects &reloads changes in the  initializer files

Reply:
I also think that we can remove some of the safety features when parsing the email. It's good that you thought about the failure cases here, but a successful call to `.to_s` should always result in a string being returned, making the following `&.` redundant.

I'd reduce it too:

```suggestion
      JSON.parse(req.body.string).dig('auth', 'email')&.to_s.downcase.gsub(/\s+/, '').presence
```

Reply:
No, and no. Thanks! I'll try to replicate your approach.

P.S. `puts "JSON::ParserError"` was just for testing, should I return `false` in the rescue block?

Reply:
I was wondering the same what the correct return value here is. Checking the example documentation https://github.com/rack/rack-attack/blob/master/docs/example_configuration.md I see `if` clauses being used, so returning `nil` implicitly or explicitly should be the way to go here in order for the throttling block to ignore that request.

Reply:
@nTraum Removing the `&` symbols results in the test raising error: `NoMethodError:undefined method `downcase' for nil:NilClass` (or similar for gsub).

Note: I did not write this line, I just copied it from similar existing rule.

Reply:
Following a call, we agreed to not remove the `&` symbols, since the protect against being passed a `nil` object, and to still expect a 500 response after rescuing JSON parse error.